{
 "metadata": {
  "name": ""
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from Bio.SeqIO import parse as parse_seq_file\n",
      "\n",
      "recs = parse_seq_file('../seq/16S.ungap.afn', 'fasta')\n",
      "\n",
      "#i = iter(recs)\n",
      "#a = i.next()\n",
      "#b = i.next()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "recs_list = list(recs)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def perc_identity(seq1, seq2):\n",
      "    \"\"\"Calculate the percent nucleotide identity between two sequences.\n",
      "    \n",
      "    Does not count shared gaps.\n",
      "    Counts unshared gaps.\n",
      "    \"\"\"\n",
      "    seq1 = asarray(seq1)\n",
      "    seq2 = asarray(seq2)\n",
      "    in_one = (seq1 != '-') | (seq2 != '-')\n",
      "    identical = (seq1 == seq2)\n",
      "    return float(sum(identical & in_one)) / sum(in_one)\n",
      "\n",
      "#%timeit(perc_identity(a, b))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def chunks(iterator, n, part_last=False):\n",
      "    out = []\n",
      "    for i, item in enumerate(iterator):\n",
      "        out.append(item)\n",
      "        if (i != 0) and (i % n):\n",
      "            yield out\n",
      "            out = []\n",
      "    if out and part_last:\n",
      "        yield out"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "copies_map = {}\n",
      "with open('../meta/16S.tsv') as handle:\n",
      "    for line in handle:\n",
      "        fields = line.split()\n",
      "        key = fields[0]\n",
      "        try:\n",
      "            value = int(fields[3])\n",
      "        except ValueError:\n",
      "            continue\n",
      "        else:\n",
      "            if key in copies_map:\n",
      "                assert copies_map[key] == value\n",
      "            else:\n",
      "                copies_map[key] = value\n",
      "                \n",
      "genome_map = {}\n",
      "with open('../meta/16S.tsv') as handle:\n",
      "    for line in handle:\n",
      "        fields = line.split()\n",
      "        key = fields[0]\n",
      "        try:\n",
      "            value = int(fields[1])\n",
      "        except ValueError:\n",
      "            continue\n",
      "        else:\n",
      "            if key in genome_map:\n",
      "                assert genome_map[key] == value\n",
      "            else:\n",
      "                genome_map[key] = value"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Takes about 10 minutes to run.\n",
      "\n",
      "import random\n",
      "from pandas import DataFrame\n",
      "\n",
      "n = 10000\n",
      "table = []\n",
      "counter = 0\n",
      "for seq1, seq2 in chunks(random.sample(recs_list, n), 2):\n",
      "    table.append((seq1.name, seq2.name,\n",
      "                  genome_map[seq1.name], genome_map[seq2.name],\n",
      "                  copies_map[seq1.name], copies_map[seq2.name],\n",
      "                  perc_identity(seq1, seq2)))\n",
      "\n",
      "\n",
      "transpose = zip(*table)\n",
      "data = DataFrame(dict(leaf1 = transpose[0],\n",
      "                      leaf2 = transpose[1],\n",
      "                      genome1 = transpose[2],\n",
      "                      genome2 = transpose[3],\n",
      "                      copy1 = transpose[4],\n",
      "                      copy2 = transpose[5],\n",
      "                      perc_ident = transpose[6]))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "hist(data.perc_ident, bins=50)\n",
      "print"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data['xwob'] = rand(len(data)) - 0.5\n",
      "data['ywob'] = rand(len(data)) - 0.5\n",
      "scatter(data.copy1 + data.xwob, data.copy2 + data.ywob,\n",
      "        s=100,\n",
      "        c=data.perc_ident, cmap='hot',\n",
      "        alpha=0.35, lw=0, marker='.')\n",
      "colorbar()\n",
      "plt.xlim(1,15)\n",
      "plt.ylim(1,15)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#data['genome1'] = data.leaf1.map(genome_map)\n",
      "#data['genome2'] = data.leaf2.map(genome_map)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import statsmodels.api as sm\n",
      "\n",
      "dists = linspace(0.5, 1, 100)\n",
      "r2s = empty_like(dists)\n",
      "coefs = empty_like(dists)\n",
      "ci_low = empty_like(dists)\n",
      "ci_hi = empty_like(dists)\n",
      "for i, d in enumerate(dists):\n",
      "    sub_data = data[data.perc_ident > d]\n",
      "    if len(sub_data) < 2:\n",
      "        r2s[i] = float(nan)\n",
      "        coefs[i] = float(nan)\n",
      "        ci_low[i] = float(nan)\n",
      "        ci_hi[i] = float(nan)\n",
      "        continue\n",
      "    lm = sm.OLS(sub_data['copy1'], sub_data['copy2'])\n",
      "    fit = lm.fit()\n",
      "    r2s[i] = fit.rsquared\n",
      "    coefs[i] = fit.params[0]\n",
      "    ci = fit.conf_int()\n",
      "    ci_low[i] = ci[0]\n",
      "    ci_hi[i] = ci[1]\n",
      "fig, ax1 = plt.subplots()\n",
      "ax1.plot(dists, r2s, label='$R^2$', lw=2)\n",
      "ax1.fill_between(dists, ci_low, ci_hi, lw=0, alpha=0.2, label=\"95% Confidence Band\", color=\"green\")\n",
      "ax1.plot(dists, coefs, label=\"$\\\\hat{\\\\beta}$\", lw=2)\n",
      "ax1.axhline(1, ls=\"--\", color=\"k\")\n",
      "ax1.set_xlabel(\"Minimum Percent Identity\")\n",
      "ax1.legend(loc=\"upper left\")\n",
      "ax1.set_xlim((min(dists), max(dists)))\n",
      "ax2 = ax1.twinx()\n",
      "ax2.hist(data.perc_ident, bins=dists, normed=True, alpha=0.15, color='k', lw=0)\n",
      "ax2.set_ylabel(\"Probability Density\")\n",
      "ax2.set_yticklabels([])\n",
      "ax2.set_xlim((min(dists), max(dists)))\n",
      "\n",
      "#fig.savefig('../res/2014-09-01_fig1.png')   # Depracated, see figure below."
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import statsmodels.api as sm\n",
      "\n",
      "window_size = 0.05\n",
      "\n",
      "dists = linspace(0.5, 1, 100)\n",
      "r2s = empty_like(dists)\n",
      "rs = empty_like(dists)\n",
      "coefs = empty_like(dists)\n",
      "ci_low = empty_like(dists)\n",
      "ci_hi = empty_like(dists)\n",
      "for i, d in enumerate(dists):\n",
      "    sub_data = data[(data.perc_ident < (d + (window_size / 2))) & \\\n",
      "                    (data.perc_ident > (d - (window_size / 2)))]  # Moving window\n",
      "    if len(sub_data) < 2:\n",
      "        r2s[i] = float(nan)\n",
      "        coefs[i] = float(nan)\n",
      "        ci_low[i] = float(nan)\n",
      "        ci_hi[i] = float(nan)\n",
      "        continue\n",
      "    lm = sm.OLS(sub_data['copy1'], sub_data['copy2'])\n",
      "    fit = lm.fit()\n",
      "    r2s[i] = fit.rsquared\n",
      "    rs[i] = corrcoef(sub_data['copy1'], sub_data['copy2'])[0,1]\n",
      "    coefs[i] = fit.params[0]\n",
      "    ci = fit.conf_int()\n",
      "    ci_low[i] = ci[0]\n",
      "    ci_hi[i] = ci[1]\n",
      "fig, ax1 = plt.subplots()\n",
      "ax1.plot(dists, r2s, label='$R^2$', lw=2)\n",
      "ax1.fill_between(dists, ci_low, ci_hi, lw=0, alpha=0.2, label=\"95% Confidence Band\", color=\"green\")\n",
      "ax1.plot(dists, coefs, label=\"$\\\\hat{\\\\beta}$\", lw=2)\n",
      "ax1.plot(dists, rs, label='$r$', lw=2)\n",
      "ax1.axhline(1, ls=\"--\", color=\"k\")\n",
      "ax1.set_xlabel(\"Minimum Percent Identity\")\n",
      "ax1.legend(loc=\"upper left\")\n",
      "ax1.set_xlim((min(dists), max(dists)))\n",
      "ax2 = ax1.twinx()\n",
      "ax2.hist(data.perc_ident, bins=dists, normed=True, alpha=0.15, color='k', lw=0)\n",
      "ax2.set_ylabel(\"Probability Density\")\n",
      "ax2.set_yticklabels([])\n",
      "ax2.set_xlim((min(dists), max(dists)))\n",
      "\n",
      "#fig.savefig('../res/2014-09-01_fig1.png')   # Depracated, see figure below."
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import random\n",
      "\n",
      "def random_combination(population, k, n=-1):\n",
      "    i = 0\n",
      "    while True:\n",
      "        yield random.sample(population, k)\n",
      "        i += 1\n",
      "        if i == n:\n",
      "            break"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pandas import DataFrame\n",
      "\n",
      "n = 100000\n",
      "table = []\n",
      "for seq1, seq2 in random_combination(recs_list, 2, n):\n",
      "    table.append((seq1.name, seq2.name,\n",
      "                  genome_map[seq1.name], genome_map[seq2.name],\n",
      "                  copies_map[seq1.name], copies_map[seq2.name],\n",
      "                  perc_identity(seq1, seq2)))\n",
      "\n",
      "\n",
      "\n",
      "transpose = zip(*table)\n",
      "data2 = DataFrame(dict(leaf1 = transpose[0],\n",
      "                       leaf2 = transpose[1],\n",
      "                       genome1 = transpose[2],\n",
      "                       genome2 = transpose[3],\n",
      "                       copy1 = transpose[4],\n",
      "                       copy2 = transpose[5],\n",
      "                       perc_ident = transpose[6]))\n",
      "data2"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data2['xwob'] = rand(len(data2)) - 0.5\n",
      "data2['ywob'] = rand(len(data2)) - 0.5\n",
      "scatter(data2.copy1 + data2.xwob, data2.copy2 + data2.ywob,\n",
      "        s=10,\n",
      "        c=data2.perc_ident, cmap='cool',\n",
      "        alpha=1, lw=0, marker='.')\n",
      "colorbar()\n",
      "plt.xlim(1,15)\n",
      "plt.ylim(1,15)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import statsmodels.api as sm\n",
      "\n",
      "window_size = 0.01\n",
      "\n",
      "dists = linspace(0.5, 1, 200)\n",
      "r2s = empty_like(dists)\n",
      "rs = empty_like(dists)\n",
      "coefs = empty_like(dists)\n",
      "ci_low = empty_like(dists)\n",
      "ci_hi = empty_like(dists)\n",
      "ssr = empty_like(dists)\n",
      "dfr = empty_like(dists)\n",
      "for i, d in enumerate(dists):\n",
      "    sub_data = data2[(data2.perc_ident < (d + (window_size / 2))) & \\\n",
      "                     (data2.perc_ident > (d - (window_size / 2)))]  # Moving window\n",
      "    if len(sub_data) < 2:\n",
      "        r2s[i] = float(nan)\n",
      "        coefs[i] = float(nan)\n",
      "        ci_low[i] = float(nan)\n",
      "        ci_hi[i] = float(nan)\n",
      "        continue\n",
      "    lm = sm.OLS(sub_data['copy1'], sub_data['copy2'])\n",
      "    fit = lm.fit()\n",
      "    r2s[i] = fit.rsquared\n",
      "    rs[i] = corrcoef(sub_data['copy1'], sub_data['copy2'])[0,1]\n",
      "    coefs[i] = fit.params[0]\n",
      "    ci = fit.conf_int()\n",
      "    ci_low[i] = ci[0]\n",
      "    ci_hi[i] = ci[1]\n",
      "    ssr[i] = fit.ssr\n",
      "    dfr[i] = fit.df_resid\n",
      "fig, ax1 = plt.subplots()\n",
      "ax1.plot(dists, r2s, label='$R^2$', lw=2)\n",
      "ax1.fill_between(dists, ci_low, ci_hi, lw=0, alpha=0.2, label=\"95% Confidence Band\", color=\"green\")\n",
      "ax1.plot(dists, coefs, label=\"$\\\\hat{\\\\beta}$\", lw=2)\n",
      "ax1.plot(dists, rs, label='$r$', lw=2)\n",
      "ax1.axhline(1, ls=\"--\", color=\"k\")\n",
      "ax1.set_xlabel(\"Minimum Percent Identity\")\n",
      "ax1.legend(loc=\"lower right\")\n",
      "ax1.set_xlim((min(dists), max(dists)))\n",
      "ax1.invert_xaxis()\n",
      "ax2 = ax1.twinx()\n",
      "ax2.hist(data2.perc_ident, bins=dists, normed=True, alpha=0.15, color='k', lw=0)\n",
      "ax2.set_ylabel(\"Probability Density\")\n",
      "ax2.set_yticklabels([])\n",
      "ax2.set_xlim((min(dists), max(dists)))\n",
      "ax2.invert_xaxis()\n",
      "\n",
      "fig.savefig('../res/2014-09-01_fig2.pdf')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fit.ssr / fit.df_resid"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fit."
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "table.predict1_low.dtype"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import statsmodels.api as sm\n",
      "from scipy.stats import t\n",
      "from pandas import DataFrame as df\n",
      "\n",
      "window_size = 0.01\n",
      "dists = linspace(0.5, 1, 200)\n",
      "columns = ['dist_center', 'dist_wind_low', 'dist_wind_hi',\n",
      "           'R2', 'r', 'beta', 'beta_ci_low', 'beta_ci_hi',\n",
      "           'ssr', 'n', 'df_resid', 'predict1', 'predict1_std']\n",
      "table = DataFrame(index=range(len(dists)), columns=columns, dtype=float64)\n",
      "table.dist_center = dists\n",
      "table.dist_wind_low = table.dist_center - window_size / 2\n",
      "table.dist_wind_hi = table.dist_center + window_size / 2\n",
      "for i in range(len(dists)):\n",
      "    sub_data = data2[(data2.perc_ident > table.dist_wind_low[i]) & \\\n",
      "                     (data2.perc_ident < table.dist_wind_hi[i])]\n",
      "    if len(sub_data) < 2:\n",
      "        continue\n",
      "    lm = sm.OLS(sub_data['copy1'], sub_data['copy2'])\n",
      "    fit = lm.fit()\n",
      "    table.R2[i] = fit.rsquared\n",
      "    table.r[i] = corrcoef(sub_data['copy1'], sub_data['copy2'])[0,1]\n",
      "    table.beta[i] = fit.params[0]\n",
      "    ci = fit.conf_int()\n",
      "    table.beta_ci_low[i] = ci[0]\n",
      "    table.beta_ci_hi[i] = ci[1]\n",
      "    table.ssr[i] = fit.ssr\n",
      "    table.n[i] = fit.nobs\n",
      "    table.df_resid[i] = fit.df_resid\n",
      "    table.predict1[i] = fit.predict(1)\n",
      "    table.predict1_std[i] = sqrt(fit.mse_resid)\n",
      "table['predict1_low'] = table.predict1 - (t.ppf(0.975, table.df_resid) * table.predict1_std)\n",
      "table['predict1_hi'] = table.predict1 + (t.ppf(0.975, table.df_resid) * table.predict1_std)\n",
      "    \n",
      "fig, ax1 = plt.subplots()\n",
      "ax1.plot(table.dist_center, table.r, label='$r$')\n",
      "ax1.plot(table.dist_center, table.predict1, label='Prediction')\n",
      "ax1.fill_between(table.dist_center, table.predict1_low, table.predict1_hi, alpha=0.2)\n",
      "ax1.fill_between(dists, ci_low, ci_hi, lw=0, alpha=0.2, label=\"95% Confidence Band\", color=\"green\")\n",
      "ax1.plot(dists, coefs, label=\"$\\\\hat{\\\\beta}$\", lw=2)\n",
      "ax1.plot(dists, rs, label='$r$', lw=2)\n",
      "ax1.axhline(1, ls=\"--\", color=\"k\")\n",
      "ax1.set_xlabel(\"Minimum Percent Identity\")\n",
      "ax1.legend(loc=\"upper right\")\n",
      "ax1.set_xlim((min(dists), max(dists)))\n",
      "ax1.invert_xaxis()\n",
      "ax2 = ax1.twinx()\n",
      "ax2.hist(data2.perc_ident, bins=dists, normed=True, alpha=0.15, color='k', lw=0)\n",
      "ax2.set_ylabel(\"Probability Density\")\n",
      "ax2.set_yticklabels([])\n",
      "ax2.set_xlim((min(dists), max(dists)))\n",
      "ax2.invert_xaxis()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "table.predict1_hi"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from scipy.stats import t\n",
      "t.ppf(0.975, list(table.df_resid))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "array(table.R2)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import statsmodels.api as sm\n",
      "from pandas import DataFrame as df\n",
      "\n",
      "dists = linspace(0.5, 1, 200)\n",
      "columns = ['dist_center', 'dist_wind_low', 'dist_wind_hi',\n",
      "           'R2', 'r', 'beta', 'beta_ci_low', 'beta_ci_hi',\n",
      "           'ssr', 'n']\n",
      "table = DataFrame(index=range(len(dists)), columns=columns)\n",
      "table"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data2.to_csv('../res/16S.dists.sample.tsv',\n",
      "             cols=['leaf1', 'leaf2', 'genome1', 'genome2',\n",
      "                   'copy1', 'copy2', 'perc_ident'],\n",
      "             sep='\\t', index=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}